plugins/net_plugin/net_plugin.cpp

```diff
@@ -3428,10 +3470,53 @@ namespace eosio {
       }
    }
 
+   // called from any thread
+   void net_plugin_impl::update_security_group(const block_state_ptr& bs) {
+      // update cache
+      //    - connection_mtx is needed
+      //
+      auto& update = bs->get_security_group_info();
+      std::shared_lock<std::shared_mutex> connection_guard(connections_mtx);
```

Comment:
This needs to be a `lock_guard` as it is modifying state.

Reply:
This is called on every `on_accepted_block` which means we now have a mutex on the critical performance path. Can we re-design this so that we don't need to acquire a mutex on every block.